### PR TITLE
Mlacko@experimental storage

### PR DESCRIPTION
**Pull request by Kyras**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ members = [
     "networking",
     "shell",
     "storage",
+    "storage_structure",
```
[the remark this answers is not in the document]
Comment (Kyras) [reply Cargo.toml]: Fixed in 21e841f

File: storage_structure/src/lane.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+type LaneDatabase<T> = dyn DatabaseWithSchema<Lane<T>> + Sync + Send;
+
+/// Lane is an way to traverse the chain.
+/// Lane is just an linked list, containing all changes between nodes.
+/// There should be multiple lanes, to be able to skip multiple nodes and traverse structure faster.
+#[derive(Clone, Debug)]
+pub struct Lane<C: ListValue> {
+    level: usize,
+    db: Arc<DB>,
```
[the remark this answers is not in the document]
Comment (Kyras) [reply storage_structure/src/lane.rs]: As per our conversation, current action should be wrapping `SkipList` in some sync-entity like `RWLock`. For future, there are more elaborate ways to achieve thread-safety. Closing for now.

File: storage_structure/src/lane.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    _pd: PhantomData<C>,
+}
+
+impl<C: ListValue> Schema for Lane<C> {
+    const COLUMN_FAMILY_NAME: &'static str = "skip_list_lanes";
```
[the remark this answers is not in the document]
Comment (Kyras) [reply storage_structure/src/lane.rs]: Such change would require change in the `Schema` trait, as only way to create `&'static str` is with `concat!` macro, which requires string literals. Changing `COLUMN_FAMILY_NAME` to a respective function is, IMHO, better solution.

File: storage_structure/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        fn diff(&mut self, other: &Self) {
+            for x in &other.0 {
+                if !self.0.contains(x) {
+                    self.0.insert(*x);
+                }
+            }
+        }
+    }
+
+    impl Codec for Value {
```
[the remark this answers is not in the document]
Comment (Kyras) [reply storage_structure/src/lib.rs]: Wouldn't it be better to implement `Codec` trait for `BincodeEncoded` traits globally ?

File: storage_structure/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+version = "0.1.0"
+authors = ["Martin Lacko <martin.lacko@simplestaking.com>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+rocksdb = "0.12"
+serde = { version = "1.0", features = ["derive"] }
+bincode = "1.2.0"
```
[the remark this answers is not in the document]
Comment (Kyras) [reply storage_structure/Cargo.toml]: Fixed in 82ef067

File: storage_structure/src/lane.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    /// From starting index, iterate backwards.
+    pub fn base_iterator(&self, starting_index: usize) -> Option<IteratorWithSchema<Lane<C>>> {
+        self.container().iterator(IteratorMode::From(
+            &NodeHeader::new(self.level, starting_index), Direction::Reverse,
+        )).ok()
+    }
+
+    fn container(&self) -> &Arc<impl DatabaseWithSchema<Lane<C>>> { &self.db }
```
[the remark this answers is not in the document]
Comment (Kyras) [reply storage_structure/src/lane.rs]: Fixed in 82ef067

File: storage_structure/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrjoe7) [reply storage_structure/src/lib.rs]: it is https://github.com/simplestaking/tezedge/blob/master/storage/src/persistent/schema.rs#L61-L69

File: storage_structure/src/lane.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrjoe7) [reply storage_structure/src/lane.rs]: I agree, function will be a better solution in this case

File: storage_structure/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kyras) [reply storage_structure/src/lib.rs]: Aaaha, makes sense

[the remark this answers is not in the document]
Comment (Kyras) [reply storage_structure/src/lib.rs]: Changed in a2a575c

File: storage_structure/src/lane.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kyras) [reply storage_structure/src/lane.rs]: This is out of the scope of this structure implementation, and should be handled in separated issue. marking this as resolved.

**Pull request merged**